File: sdk/typescript/src/adpkg.ts
import fs from "fs";
import path from "path";
import * as tar from "tar";
import crypto from "crypto";
import yaml from "js-yaml";
import { validateAdp } from "./validation.js";

const OCI_LAYOUT = { imageLayoutVersion: "1.0.0" };
const MANIFEST_MEDIA_TYPE = "application/vnd.oci.image.manifest.v1+json";
const LAYER_MEDIA_TYPE = "application/vnd.adp.package.v1+tar";
const CONFIG_MEDIA_TYPE = "application/vnd.adp.config.v1+json";

function sha256(buf: Buffer): string {
  return "sha256:" + crypto.createHash("sha256").update(buf).digest("hex");
}

function writeBlob(blobsDir: string, digest: string, data: Buffer) {
  const [, hex] = digest.split(":");
  const dir = path.join(blobsDir, "sha256");
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(path.join(dir, hex), data);
  return data.length;
}

export function createPackage(
  srcDir: string,
  outDir: string,
  builderId = "",
  sourceRepo = "",
  sourceRef = "",
) {
  const adpPath = path.join(srcDir, "adp", "agent.yaml");
  const adpContent = fs.readFileSync(adpPath, "utf8");
  const adpObj = yaml.load(adpContent) as any;
  const errors = validateAdp(adpObj);
  if (errors.length) throw new Error(`ADP validation failed: ${errors.join("; ")}`);

  fs.mkdirSync(outDir, { recursive: true });
  const blobsRoot = path.join(outDir, "blobs");
  fs.mkdirSync(path.join(blobsRoot, "sha256"), { recursive: true });

  // Config blob with provenance fields
  const configBuf = Buffer.from(JSON.stringify({
    agent_id: adpObj.id,
    adp_version: adpObj.adp_version,
    "builder.id": builderId,
    "source.repo": sourceRepo,
    "source.ref": sourceRef,
    build_timestamp: new Date().toISOString(),
  }));
  const configDigest = sha256(configBuf);
  const configSize = writeBlob(blobsRoot, configDigest, configBuf);

  // Layer tar
  const layerTar = path.join(outDir, "layer.tar");
  tar.c({ file: layerTar, cwd: srcDir, sync: true }, fs.readdirSync(srcDir));
  const layerBuf = fs.readFileSync(layerTar);
  const layerDigest = sha256(layerBuf);
  const layerSize = writeBlob(blobsRoot, layerDigest, layerBuf);
  fs.unlinkSync(layerTar);

  const manifest = {
    schemaVersion: 2,
    mediaType: MANIFEST_MEDIA_TYPE,
    config: { mediaType: CONFIG_MEDIA_TYPE, digest: configDigest, size: configSize },
    layers: [{ mediaType: LAYER_MEDIA_TYPE, digest: layerDigest, size: layerSize }],
  };
  const manifestBuf = Buffer.from(JSON.stringify(manifest, null, 2));
  const manifestDigest = sha256(manifestBuf);
  writeBlob(blobsRoot, manifestDigest, manifestBuf);

  const index = {
    schemaVersion: 2,
    manifests: [
      {
        mediaType: MANIFEST_MEDIA_TYPE,
        digest: manifestDigest,
        size: manifestBuf.length,
        annotations: { "org.opencontainers.image.title": adpObj.id },
      },
    ],
  };
  fs.writeFileSync(path.join(outDir, "index.json"), JSON.stringify(index, null, 2));
  fs.writeFileSync(path.join(outDir, "oci-layout"), JSON.stringify(OCI_LAYOUT));
}

function blobPath(pkgDir: string, digest: string): string {
  return path.join(pkgDir, "blobs", digest.replace("sha256:", "sha256/"));
}

export function inspectPackage(pkgDir: string): { agent_id: string; adp_version: string; layer_count: number; config: Record<string, unknown> } {
  const index = JSON.parse(fs.readFileSync(path.join(pkgDir, "index.json"), "utf8"));
  const manifestDesc = index.manifests[0];
  const manifest = JSON.parse(fs.readFileSync(blobPath(pkgDir, manifestDesc.digest), "utf8"));
  const config = JSON.parse(fs.readFileSync(blobPath(pkgDir, manifest.config.digest), "utf8"));
  return {
    agent_id: config.agent_id,
    adp_version: config.adp_version,
    layer_count: manifest.layers.length,
    config,
  };
}

export function verifyPackage(pkgDir: string): { passed: boolean; failures: string[] } {
  const failures: string[] = [];
  const index = JSON.parse(fs.readFileSync(path.join(pkgDir, "index.json"), "utf8"));
  for (const entry of index.manifests ?? []) {
    verifyBlob(pkgDir, entry.digest, failures);
    const manifest = JSON.parse(fs.readFileSync(blobPath(pkgDir, entry.digest), "utf8"));
    if (manifest.config?.digest) verifyBlob(pkgDir, manifest.config.digest, failures);
    /* c8 ignore next */
    for (const layer of manifest.layers ?? []) {
      if (layer.digest) verifyBlob(pkgDir, layer.digest, failures);
    }
  }
  return { passed: failures.length === 0, failures };
}

function verifyBlob(pkgDir: string, digest: string, failures: string[]): void {
  const bp = blobPath(pkgDir, digest);
  if (!fs.existsSync(bp)) { failures.push(`${digest}: blob file missing`); return; }
  const actual = sha256(fs.readFileSync(bp));
  if (actual !== digest) failures.push(`${digest}: digest mismatch (actual ${actual})`);
}

export function openPackage(pkgDir: string) {
  const index = JSON.parse(fs.readFileSync(path.join(pkgDir, "index.json"), "utf8"));
  const manifestDesc = index.manifests[0];
  const manifestPath = path.join(pkgDir, "blobs", manifestDesc.digest.replace("sha256:", "sha256/"));
  const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
  const layerDesc = manifest.layers[0];
  const layerPath = path.join(pkgDir, "blobs", layerDesc.digest.replace("sha256:", "sha256/"));
  const extractDir = fs.mkdtempSync(path.join(pkgDir, "extract-"));
  tar.x({ file: layerPath, cwd: extractDir, sync: true });
  const adpContent = fs.readFileSync(path.join(extractDir, "adp", "agent.yaml"), "utf8");
  return yaml.load(adpContent) as any;
}

File: sdk/python/adp_sdk/adpkg.py
from __future__ import annotations

import hashlib
import json
import tarfile
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable, List

import yaml

from .adp_model import ADP
from .validation import validate_adp

OCI_LAYOUT = {"imageLayoutVersion": "1.0.0"}
MANIFEST_MEDIA_TYPE = "application/vnd.oci.image.manifest.v1+json"
LAYER_MEDIA_TYPE = "application/vnd.adp.package.v1+tar"
CONFIG_MEDIA_TYPE = "application/vnd.adp.config.v1+json"


@dataclass
class Descriptor:
    mediaType: str
    digest: str
    size: int
    annotations: dict | None = None

    def to_dict(self) -> dict:
        data = {
            "mediaType": self.mediaType,
            "digest": self.digest,
            "size": self.size,
        }
        if self.annotations:
            data["annotations"] = self.annotations
        return data


class ADPackage:
    """OCI-based ADP package helper."""

    def __init__(self, path: Path):
        self.path = Path(path)

    @staticmethod
    def _iter_files(root: Path) -> Iterable[Path]:
        for path in root.rglob("*"):
            if path.is_file():
                yield path

    @staticmethod
    def _blob_path(blobs: Path, digest: str) -> Path:
        algo, hexval = digest.split(":", 1)
        return blobs / algo / hexval

    @staticmethod
    def _hash_file(path: Path) -> tuple[str, int]:
        hasher = hashlib.sha256()
        size = 0
        with path.open("rb") as f:
            for chunk in iter(lambda: f.read(8192), b""):
                hasher.update(chunk)
                size += len(chunk)
        return f"sha256:{hasher.hexdigest()}", size

    @staticmethod
    def _hash_bytes(data: bytes) -> tuple[str, int]:
        h = hashlib.sha256(data).hexdigest()
        return f"sha256:{h}", len(data)

    @classmethod
    def create_from_directory(
        cls,
        src: str | Path,
        out_path: str | Path,
        builder_id: str = "",
        source_repo: str = "",
        source_ref: str = "",
    ) -> "ADPackage":
        src_path = Path(src)
        out_dir = Path(out_path)
        out_dir.mkdir(parents=True, exist_ok=True)
        if out_dir.suffix != "":
            raise ValueError(
                "OCI layout is a directory; provide a directory path, not a file"
            )

        adp_path = src_path / "adp" / "agent.yaml"
        adp = ADP.from_file(adp_path)
        validate_adp(adp)

        blobs = out_dir / "blobs" / "sha256"
        blobs.mkdir(parents=True, exist_ok=True)

        # Config blob with provenance fields
        config_bytes = json.dumps({
            "agent_id": adp.id,
            "adp_version": adp.adp_version,
            "builder.id": builder_id,
            "source.repo": source_repo,
            "source.ref": source_ref,
            "build_timestamp": datetime.now(timezone.utc).isoformat(),
        }).encode()
        config_digest, config_size = cls._hash_bytes(config_bytes)
        config_path = cls._blob_path(out_dir / "blobs", config_digest)
        config_path.parent.mkdir(parents=True, exist_ok=True)
        config_path.write_bytes(config_bytes)
        config_desc = Descriptor(CONFIG_MEDIA_TYPE, config_digest, config_size)

        # Layer blob: tar of src directory contents
        layer_tar = out_dir / "layer.tar"
        with tarfile.open(layer_tar, "w") as tar:
            for file_path in cls._iter_files(src_path):
                arcname = file_path.relative_to(src_path)
                tar.add(file_path, arcname=arcname)
        layer_digest, layer_size = cls._hash_file(layer_tar)
        layer_blob_path = cls._blob_path(out_dir / "blobs", layer_digest)
        layer_blob_path.parent.mkdir(parents=True, exist_ok=True)
        layer_blob_path.write_bytes(layer_tar.read_bytes())
        layer_tar.unlink()
        layer_desc = Descriptor(LAYER_MEDIA_TYPE, layer_digest, layer_size)

        manifest = {
            "schemaVersion": 2,
            "mediaType": MANIFEST_MEDIA_TYPE,
            "config": config_desc.to_dict(),
            "layers": [layer_desc.to_dict()],
        }
        manifest_bytes = json.dumps(manifest, indent=2).encode()
        manifest_digest, manifest_size = cls._hash_bytes(manifest_bytes)
        manifest_path = cls._blob_path(out_dir / "blobs", manifest_digest)
        manifest_path.parent.mkdir(parents=True, exist_ok=True)
        manifest_path.write_bytes(manifest_bytes)

        index = {
            "schemaVersion": 2,
            "manifests": [
                {
                    "mediaType": MANIFEST_MEDIA_TYPE,
                    "digest": manifest_digest,
                    "size": manifest_size,
                    "annotations": {"org.opencontainers.image.title": adp.id},
                }
            ],
        }
        (out_dir / "index.json").write_text(json.dumps(index, indent=2))
        (out_dir / "oci-layout").write_text(json.dumps(OCI_LAYOUT))
        return cls(out_dir)

    @classmethod
    def open(cls, path: str | Path) -> "ADPackage":
        return cls(Path(path))

    def inspect(self) -> dict:
        """Return structured metadata: agent_id, adp_version, layer count, config blob contents."""
        index = json.loads((self.path / "index.json").read_text())
        manifest_desc = index["manifests"][0]
        manifest_bytes = self._blob_path(self.path / "blobs", manifest_desc["digest"]).read_bytes()
        manifest = json.loads(manifest_bytes)
        config_bytes = self._blob_path(self.path / "blobs", manifest["config"]["digest"]).read_bytes()
        config = json.loads(config_bytes)
        return {
            "agent_id": config.get("agent_id"),
            "adp_version": config.get("adp_version"),
            "layer_count": len(manifest.get("layers", [])),
            "config": config,
        }

    def verify(self) -> dict:
        """Recompute SHA-256 of each blob and compare to stored digest. Returns pass/fail + failures."""
        failures: List[str] = []
        index = json.loads((self.path / "index.json").read_text())
        for entry in index.get("manifests", []):
            self._verify_blob(entry["digest"], failures)
            blob_path = self._blob_path(self.path / "blobs", entry["digest"])
            if not blob_path.exists():
                continue
            manifest = json.loads(blob_path.read_bytes())
            config_desc = manifest.get("config", {})
            if config_desc.get("digest"):
                self._verify_blob(config_desc["digest"], failures)
            for layer in manifest.get("layers", []):
                if layer.get("digest"):
                    self._verify_blob(layer["digest"], failures)
        return {"passed": len(failures) == 0, "failures": failures}

    def _verify_blob(self, digest: str, failures: List[str]) -> None:
        blob_path = self._blob_path(self.path / "blobs", digest)
        if not blob_path.exists():
            failures.append(f"{digest}: blob file missing")
            return
        actual_digest, _ = self._hash_bytes(blob_path.read_bytes())
        if actual_digest != digest:
            failures.append(f"{digest}: digest mismatch (actual {actual_digest})")

    def list_blobs(self) -> List[str]:
        return [p.name for p in (self.path / "blobs" / "sha256").glob("*")]

    def read_adp(self) -> ADP:
        # Extract layer tar and read adp/agent.yaml
        index = json.loads((self.path / "index.json").read_text())
        manifest_desc = index["manifests"][0]
        manifest = json.loads(
            (
                self.path
                / "blobs"
                / manifest_desc["digest"].replace("sha256:", "sha256/")
            ).read_text()
        )
        layer_desc = manifest["layers"][0]
        layer_path = (
            self.path / "blobs" / layer_desc["digest"].replace("sha256:", "sha256/")
        )
        with tarfile.open(layer_path, "r") as tar:
            try:
                member = tar.extractfile("adp/agent.yaml")
            except KeyError:
                raise FileNotFoundError("adp/agent.yaml not found in layer")
            if not member:
                raise FileNotFoundError("adp/agent.yaml not found in layer")
            data = member.read().decode()
        return ADP.model_validate(yaml.safe_load(data))
